src/libaktualizr/aktualizr_api.h

```diff
@@ -0,0 +1,241 @@
+  /** The device is not provisioned, and provisioning isn't possible and/or wasn't requested */
+  AKTUALIZR_NOT_PROVISIONED,
+  /** Network communication failed. For example a timeout, DNS failure, no network is present. */
+  AKTUALIZR_NETWORK_ERROR,
+
+} aktualizr_res_t;
+
+/**
+ * High-level update state suitable for display to a user.
+ * This interface is expected to be used by a UI, therefore transient errors
+ * like network failure are not exposed here. The system--not the user--will be
+ * responsible for retrying in those cases.
+ */
+typedef enum {
+  /** The system is up-to-date with respect to the most recent metadata downloaded. */
+  AKTUALIZR_UPDATE_UP_TO_DATE,
+  /**
+   * A metadata fetch is in progress.
+   * A UI may display this identically to \ref AKTUALIZR_UPDATE_UP_TO_DATE, but
+   * the state transitions can be used to update a 'last checked for updates 1
+   * hour ago' UI component.
+   */
+  AKTUALIZR_UPDATE_CHECK_IN_PROGRESS,
+
+  /**
+   * An update is available.
+   */
+  AKTUALIZR_UPDATE_AVAILABLE,
+
+  /**
+   * An update is downloading
+   */
+  AKTUALIZR_UPDATE_DOWNLOADING,
+
+  /**
+   * An update is installing
+   */
+  AKTUALIZR_UPDATE_INSTALLING,
+
+  /**
+   * An update is already installed, but a reboot is required to start using it.
+   */
+  AKTUALIZR_UPDATE_AWAITING_REBOOT
+} aktualizr_update_state_t;
+
+/**
+ * Parameter to \ref aktualizr_run_update to control how much of the update
+ * cycle to perform.
+ */
+typedef enum {
+  /** Perform a full update cycle */
+  AKTUALIZR_UPDATE_FULL_CYCLE = 1,
+  /** Fetch Update metadata only */
+  AKTUALIZR_UPDATE_METADATA,
+  /** Download only */
+  AKTUALIZR_UPDATE_DOWNLOAD,
+  /** Install only */
+  AKTUALIZR_UPDATE_INSTALL
+
+} aktualizr_poll_operation_t;
+
+/** Callback type for \ref aktualizr_set_state_change_callback */
+typedef void (*aktualizr_callback_t)(void* clientp, aktualizr_update_state_t);
+
+/// \name Main API
+/// \{
+
+/**
+ * Create a new Aktualizr software updater.
+ * If polling is enabled in the configuration, this will start polling the
+ * update server in the background.
+ */
+aktualizr_t aktualizr_init(aktualizr_config_t configuration);
+
+/**
+ * Shutdown Aktualizr. No other calls to this API may be made after this call.
+ */
+void aktualizr_free(aktualizr_t aktualizr);
+
+/**
+ * Register to be notified when updates are available, downloading or installing.
+ * It is not possible to register to multiple callbacks. Passing NULL as a
+ * callback will disable state change notifications.
+ * \param aktualizr Aktualizr handle
+ * \param callback The callback to call, or 'NULL' to disable notifications
+ * \param userdata Opaque user handle that is passed to the callback
+ */
+void aktualizr_set_state_change_callback(aktualizr_t aktualizr, aktualizr_callback_t callback, void* userdata);
+
+/**
+ * The current state of the update process.
+ */
+aktualizr_update_state_t aktualizr_get_update_state(aktualizr_t aktualizr);
```

Comment:
Maybe this should also return a "status" value somehow. eg
~~~
 aktualizr_update_state_t aktualizr_get_update_state(aktualizr_t aktualizr, char **state_status);
~~~
Where you could then do something like:
~~~
aktualizr_update_state_t state;
char *status=NULL
state = aktualizr_get_update_state(aktualizr, &status);
if (state == AKTUALIZR_UPDATE_DOWNLOADING && status) {
    printf("Downloading update. %s\n", status);
    free(status);
}
~~~

Reply:
That's a reasonable idea. Free-ing a pointer allocated on the other side of an API boundary is a bit dangerous (we'd have to expose a `aktualizr_free()` call).

Maybe the API should be:

    void aktualizr_get_update_state(aktualizr_t, aktualizr_update_state_t* char**);

Reply:
I like that better.

Reply:
Thinking some more about my proposed API, I'm a bit worried about _just_ having a text string for the information.  It seems to me like we might need a 3rd piece of information soon, and adding a `aktualizr_get_update_icon` call would open up race conditions between the calls, because the aktualizr engine is running in a background thread.  

Maybe `aktualizr_update_state_t` needs to be a full blown object that represents a snapshot of the current state, and can be queried at will.

Reply:
@cajun-rat is it `void aktualizr_get_update_state(aktualizr_t, aktualizr_update_state_t*, char**);`? I don't understand how it would solve the allocation problem.

Alternatively, `aktualizr_update_state_t` could be a struct with a `char status[MAX_STATUS_LENGTH]` member so that heap allocation can be avoided entirely (with the signature `void aktualizr_get_update_state(aktualizr_t, aktualizr_update_state_t*)` )

Reply:
I like the `char status[MAX_STATUS_LENGTH]` approach. Then a user could do something really simple like:
~~~
aktualizr_update_state_t update_state;

/* repeatable safe calls to */
aktualizr_get_update_state(aktualizr, &update_state)
~~~

Reply:
My worry there was with ABI compatibility. If we went down that route then the size of `aktualizr_update_state_t` would be fixed. We can include a handful of padding fields that can be re-purposed for small changes, and define aktualizr_get_update_state_ex in the future to support more detailed information if needed.

Making it much harder for users to leak memory is certainly a valuable benefit of this approach.

Reply:
Maybe we are looking at this the wrong way. Polling for update status is actually a bit of a pain anyway. The caller's going to need some potentially thread-based logic to try and understand what's going on and will wind up polling at some random rate. Couldn't we get rid of this API function, and just make the `void aktualizr_set_state_change_callback(aktualizr_t aktualizr, aktualizr_callback_t callback, void* userdata);` API include this kind of additional information?

Reply:
The idea for a polling/query interface was to allow simple command line applications or for UI initialization, and I'd expect that most applications would use `aktualizr_set_state_change_callback` rather than poll `aktualizr_get_update_state`. If the polling interface can be emulated with the callback interface, then that would be a good argument for dropping it.

In terms of what data is available, the callback should receive the same information that can be fetched from  `aktualizr_get_update_state`:

    typedef void (*aktualizr_callback_t)(void* clientp, aktualizr_update_state_t);
    void aktualizr_set_state_change_callback(aktualizr_t aktualizr, aktualizr_callback_t callback, void* userdata);

(In the last update I made `aktualizr_update_state_t` an object rather than a enum)

Reply:
The latest version works for me so ignore the rest of my comment if you want :) I was just thinking removing the polling interface might be nice because you wouldn't have to worry about struct sizes. ie - if there's just a callback, then the aktualizr engine owns the memory in the struct and the user has nothing to worry about (unless they want to *copy* data from it). I'm not totally sure about ABI compatibility, but I think you could also then just append members to the struct w/o fear of breaking old code.